keps/sig-network/4631-loadbalancerstatus/README.md

```diff
@@ -0,0 +1,1066 @@
+   requested semantics.
+
+3. Define new API to allow the user to specify when they would prefer
+   that the cloud fail to provision a load balancer rather than
+   partially provisioning it.
+
+### Risks and Mitigations
+
+Using the new features should not introduce any new risks.
+
+They should hopefully _improve_ security for users. Currently, you
+sometimes get load balancers that aren't what you wanted (e.g.,
+failing to implement source ranges or session affinity). This KEP
+makes it possible for users to prevent this from happening.
+
+(If a cloud provider implemented the API incorrectly and claimed
+support for features that it didn't actually support, then that could
+lead to users being more confident when deploying broken load
+balancers than they are now, but the cloud provider would fail the e2e
+tests in that case, so this shouldn't happen.)
+
+## Design Details
+
+### `Service.Status.Conditions`
+
+The problem of indicating provisioning progress and provisioning
+failure can easily be solved with conditions on the Service.
+
+We will add three new Service conditions:
+
+  - `LoadBalancerServing`, to indicate explicitly that the load
+    balancer is (or isn't) serving traffic.
+
+  - `LoadBalancerProvisioning`, to indicate that the cloud provider
+    is (or isn't) configuring/provisioning the load balancer.
+
+  - `LoadBalancerDegraded`, to indicate that the load balancer is
+    aware that it is not fully implementing the semantics requested by
+    the user, in cases where implementing the full semantics is not
+    mandatory.
+
+#### The `LoadBalancerServing` Condition
+
+The `LoadBalancerServing` condition, when `True`, indicates that
+the service's load balancer is handling connections via the IP(s) or
+hostname(s) indicated in `.Status.LoadBalancer`.
+
+The cloud provider should set a `LoadBalancerServing` condition
+immediately upon first becoming aware of a `Type: LoadBalancer`
+Service. At all times, the condition should be `True` if the load
+balancer is handling connections, and `False` if it is not. (There are
+no defined semantics for an `Unknown` value of this condition.)
+
+(This means that at construction time, the `LoadBalancerServing`
+condition is redundant with the `.Status.LoadBalancer` field; it
+should switch from `False` to `True` at exactly the same time as the
+`.Status.LoadBalancer` field is filled in. However, if the Service is
+updated later, the cloud provider may temporarily switch it to
+`LoadBalancerServing=False` while reprovisioning the Service, and it
+would not generally clear `.Status.LoadBalancer` during this time.)
+
+Any time the load balancer behavior changes in an externally-visible
+way, the cloud provider must update at least the `LastTransitionTime`
+of the `LoadBalancerServing` condition. (For example, if the user
+changes the service's `LoadBalancerSourceRanges`, and the cloud
+provider updates the load balancer for this, it must update the
```

Comment:
This is not generally a requirement of Conditions - I get why that might be valuable information but it seems wrong to overload "LastTransitionTime" to mean anything but "the last time this changed state"

Reply:
ok